### ISSUE TITLE
'loaded' hook breaks order filter

### ISSUE THREAD

**Issue by jmwohl** [strongloop/loopback-datasource-juggler#861]
Might be specific to mongodb connector, doesn't seem to affect memory data source (haven't tried others). Adding a computed field via the 'loaded' hook undoes the order filter.

The following works by checking if the logged-in user has 'liked' the artwork in being loaded (via hasAndBelongsToMany relationship), and the `liked` property is added to each Artwork object as expected. However, when the results are finally returned, the order filter has no effect. Removing the ` user.liked_artwork.exists(ctx.instance.id...` query within this hook fixes the issue, but that query is necessary in order to determine the proper value for the computed property.

```
    // Add a computed 'liked' value to each artwork object at runtime
    Artwork.observe('loaded', function(ctx, next) {
        // We don't act on new instances
        if (!ctx.instance) {
            return next();
        }

        var context = loopback.getCurrentContext(),
            req = context && context.active ? context.active.http.req : null,
            user = req ? req.user : null;

        ctx.instance.liked = false;

        if (user) {
            user.liked_artwork.exists(ctx.instance.id, function(err, exists) {
                if (exists) {
                    ctx.instance.liked = true;
                }
                next();
            });
        } else {
            next();
        }
    });
```

**Comment by jacobheun** [comment]
Hitting this issue as well, it's not an issue with any particular datasource (I'm using mysql). The issue exists in the allCb function of the dao: https://github.com/strongloop/loopback-datasource-juggler/blob/v2.45.2/lib/dao.js#L1494.

Since all of the objects are being run in parallel there's no guarantee that the objects get pushed onto the result at the same time: https://github.com/strongloop/loopback-datasource-juggler/blob/v2.45.2/lib/dao.js#L1544.

The data is being loaded out of the datasource in the appropriate order, but loses the order prior to being returned from the server. Either the function needs to be re-worked, or a quick sort prior to the final callback needs to be done to ensure that the order or objects in *results* matches the original order of *data*. 

I'd be happy to submit a patch for this pending the direction the strongloop team would prefer to go with this.

**Comment by Amir-61** [comment]
I believe this is a duplicate of: https://github.com/strongloop/loopback-datasource-juggler/issues/800
Closing this ticket; however please feel free if you think it is a different issue.

PR which is pending to solve the problem: https://github.com/strongloop/loopback-datasource-juggler/pull/864

Thanks!
